File: snippets/cpp/math/matrix.cpp
using Mat = vector<vector<ll>>;
Mat matmul(Mat &a, Mat &b, ll m = MOD) {
    int n = sza(a), p = sza(b[0]);
    Mat c(n, vector<ll>(p));
    for (int i = 0; i < n; i++)
        for (int k = 0; k < sza(b); k++)
            for (int j = 0; j < p; j++)
                c[i][j] = (c[i][j] + a[i][k] * b[k][j]) % m;
    return c;
}
Mat matpow(Mat a, ll n, ll m = MOD) {
    int sz = sza(a);
    Mat r(sz, vector<ll>(sz));
    for (int i = 0; i < sz; i++) r[i][i] = 1;
    for (; n; n >>= 1, a = matmul(a, a, m))
        if (n & 1) r = matmul(r, a, m);
    return r;
}

File: snippets/rust/math/matrix.rs
type Mat = Vec<Vec<u64>>;

fn matmul(a: &Mat, b: &Mat, m: u64) -> Mat {
    let n = a.len();
    let k = b.len();
    let p = b[0].len();
    let mut c = vec![vec![0u64; p]; n];
    for i in 0..n {
        for kk in 0..k {
            if a[i][kk] == 0 { continue; }
            let aik = a[i][kk];
            for j in 0..p {
                c[i][j] = (c[i][j] + aik * b[kk][j]) % m;
            }
        }
    }
    c
}
fn matpow(mut a: Mat, mut n: u64, m: u64) -> Mat {
    let sz = a.len();
    let mut r: Mat = (0..sz).map(|i| {
        let mut row = vec![0u64; sz];
        row[i] = 1;
        row
    }).collect();
    while n > 0 {
        if n & 1 == 1 { r = matmul(&r, &a, m); }
        a = matmul(&a, &a, m);
        n >>= 1;
    }
    r
}
